accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+    npm_cli->>npm_registry: Publish package and upload Rekor entry
+    npm_registry->>fulcio: Request signing certificate
+    fulcio->>npm_registry: Signing certificate
+    npm_registry->>rekor: Upload signed release attestations
+    rekor->>npm_registry: Rekor entry
+    npm_registry->>npm_registry: Attach rekor entry to published package
+    npm_registry->>npm_cli: Success
+```
+
+1. `npm publish --build-provenance` detects execution within a supported CI/CD environment and initializes token exchange to get a JWT id token from the CI/CD system's OIDC identity provider.
+2. npm CLI creates a new ephemeral key pair, requests a certificate from Fulcio (CA) and authenticates via the JWT id token.
+3. A SLSA build provenance attestation is created based on information available within the JWT and CI system and signs it with the ephemeral key.
+4. npm CLI uploads the completed attestation record on Rekor for public packages.
+5. npm registry creates a release attestation, stating that the version has been accepted and published, and places the record on Rekor for public packages.
+
+#### Verifying
+The package integrity captured by the provenance attestation will be verified by the npm CLI. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.
+
+This will be staged rollout over several npm CLI releases as changing the default behavior of `npm install` is a breaking change:
+
+1. The existing `npm audit signatures` command will start verifying build signatures, possibly adding an opt-in flag to audit these signatures during `npm install`, e.g. `--audit-build-provenance`
+2. Start auditing build signatures by default when running `npm install`. This will only check packages that have build attestations and ignore the rest. Allow opting out of this behavior with `--no-audit-build-provenance` or `--no-audit`.
+3. Allow users to opt-in to a strict mode to require packages to be published with build provenance information, e.g. `npm install --require-build-provenance`. Initially, this might only be usable by maintainers that only depend on their dependencies. It will take a long time for this to be feasible for the average install as all transitive dependencies would need to include build provenance on the installed version. We could allow policies that make gradual enforcement possible, e.g. only require build provenance on packages published after some minimum date.
+
+Signature verification needs to be performant and have a negligible impact on `npm install`. We believe the main bottleneck here will be the network, downloading the signatures to verify. The crypto operations involved in verification are negligible on modern hardware. We propose the CLI makes a bulk API request to the npm registry, similar to how advisories are fetched, for all provenance attestations for a given install.
```

Comment:
Committing to a benchmark for this to be able to spot regressions would be great as well 👀

Reply:
Yes I like this idea 👍